### PR TITLE
Harry potter reader

### PR DESCRIPTION
**Pull request by SarahLammert**
Pls look Carl 
I will look at your comments... :)

### CODE REVIEW

File: src/harryPotterReader/Word.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package harryPotterReader;
+
+public class Word {
+	
+	private String word;
+	private int count = 1;
```
Comment (carlguo866) [inline src/harryPotterReader/Word.java]: put the `count = 1` in the constructor. This works but people don't like it as sometimes it messes stuff up. I won't go into details bc it is complicated to explain. (YOU WOULD GET POINTS OFF BC OF SOMETHING LIKE THIS ON THE AP TEST)

File: src/harryPotterReader/Word.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package harryPotterReader;
+
+public class Word {
+	
+	private String word;
```
Comment (carlguo866) [inline src/harryPotterReader/Word.java:5]: GOOD PRIVATIZATION HABITS

File: src/harryPotterReader/Word.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	
+	public Word(String s) {
+		word = s;
+	}
+	
+	public void addCount() {
+		count += 1;
+	}
+
+	public String getWord() {
```
Comment (carlguo866) [inline src/harryPotterReader/Word.java:17]: GOOD AP HABIT THAT WOULD GET YOU POINTS ON AP

I, personally, think it is such a redundant thing, but it is probably just me.

File: src/harryPotterReader/WordCounter.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
+package harryPotterReader;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class WordCounter {
+	
+	//ArrayList to hold all unique words
+	private ArrayList<Word> uniqueWords = new ArrayList<Word>(80000);
```
Comment (carlguo866) [inline src/harryPotterReader/WordCounter.java]: why 80000

File: src/harryPotterReader/WordCounter.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+			}
+			s.close();
+		} catch (Exception e) {
+			System.out.println("Problem opening file");
+		}
+		
+	}
+	
+	public static String stripSpecialChars(String w) {
+		String newW = w;
```
Comment (carlguo866) [inline src/harryPotterReader/WordCounter.java]: just make modification on the `String w` you don't need the `String newW`

File: src/harryPotterReader/WordCounter.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		newW = newW.toLowerCase();
+		
+		//Finds the first letter and places the start of the substring there
+		//Then finds the last letter
+		if(newW.length() > 1) {
+			int s = 0;
+			int f = 1;
+			boolean first = true;
+			for(int i = 0; i < newW.length(); i++) {
+				if(Character.isLetter(newW.charAt(i))) {
```
Comment (carlguo866) [inline src/harryPotterReader/WordCounter.java]: good usage of `Character.isLetter()`
You should also know stuff about the ASCII tables, in which a lower case letter would be in the range of 97 to 122. 
Reference: http://www.asciitable.com/

File: src/harryPotterReader/WordCounter.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+			//Scans the file
+			Scanner s = new Scanner(new File(fileName + ".txt"));
+			System.out.println("File is open");
+			
+			//Reads each string
+			while(s.hasNext()) {
+				String newWord = s.next();
+				
+				//Strips special characters from the string
+				newWord = stripSpecialChars(newWord);
```
Comment (carlguo866) [inline src/harryPotterReader/WordCounter.java:29]: good encapsulation here

File: src/harryPotterReader/WordCounter.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	
+	public static String stripSpecialChars(String w) {
+		String newW = w;
+		newW = newW.toLowerCase();
+		
+		//Finds the first letter and places the start of the substring there
+		//Then finds the last letter
+		if(newW.length() > 1) {
+			int s = 0;
```
Comment (carlguo866) [inline src/harryPotterReader/WordCounter.java]: good implementation. But pls write out your variable names. I couldn't really tell if `s` stands for "start" or `f` stands for "first"

File: src/harryPotterReader/WordCounter.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+			newW = newW.substring(s, f+1);
+		}
+		
+		return newW;
+	}
+	
+	public boolean isUnique(String aW) {
+		boolean numIsUnique = true;
+		for(Word w: uniqueWords) {
+			if(w.getWord().equals(aW)) {
```
Comment (carlguo866) [inline src/harryPotterReader/WordCounter.java:72]: good code here.
You can also think of a way to do things using `ArrayList.contains()`.  IN that way, you need to make modifications of the `equal()` method in the `Word` object

File: src/harryPotterReader/WordCounter.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+			}
+		}
+		return numIsUnique;
+	}
+
+	public void printAllWords() {
+		for(Word w: uniqueWords) {
+			System.out.println(w.getWord() + ": " + w.getCount());
+		}
+	}
```
Comment (carlguo866) [inline src/harryPotterReader/WordCounter.java]: instead of this, write a `String toString()` method in the Word class

**Review by carlguo866 — CHANGES_REQUESTED** [review]

**Comment by SarahLammert** [comment]
Can you look over it again? I recommitted another version...

File: src/harryPotterReader/WordCounter.java [hunk 11]
```diff
@@ -41,30 +41,29 @@ public void openFile(String fileName) {
...
+						endIndex = i;
 					}
 				}
 			}
-			newW = newW.substring(s, f+1);
+			w = w.substring(startIndex, endIndex+1);
 		}
 		
-		return newW;
+		return w;
```
Comment (carlguo866) [inline src/harryPotterReader/WordCounter.java:66]: great that you fixed everything based on my comments! ty so much

File: src/harryPotterReader/WordCounter.java [hunk 12]
```diff
@@ -81,7 +80,7 @@ public boolean isUnique(String aW) {
 
 	public void printAllWords() {
 		for(Word w: uniqueWords) {
-			System.out.println(w.getWord() + ": " + w.getCount());
+			System.out.println(w.toString());
 		}
```
Comment (carlguo866) [inline src/harryPotterReader/WordCounter.java]: great but you can just do `System.out.print(uniqueWords)`!

**Review by carlguo866 — APPROVED** [review]

**Pull request merged**